src/InverseKinematicsIPOPT.cpp

```diff
@@ -0,0 +1,499 @@
+}
+
+bool InverseKinematicsIPOPT::eval_g(Ipopt::Index n, const Number* x, bool new_x, Ipopt::Index m, Number* g)
+{
+    Transform p_H_e;
+    Eigen::Map< const Eigen::VectorXd > x_in(x, totalDOF);
+    Eigen::Map< Eigen::VectorXd > g_in(g, 8);
+    VectorDynSize joints(totalDOF-7);
+    
+    toEigen(joints) = x_in.tail(totalDOF-7);
+    iKDC.setJointPos(joints);
+    p_H_e = iKDC.getRelativeTransform(parentFrame,endEffectorFrame);
+    
+    g_in.head<3>() = toEigen(p_H_e.getPosition()) - x_in.head<3>();
+    
+    g_in.segment<4>(3) = toEigen(p_H_e.getRotation().asQuaternion()) - x_in.segment<4>(3);
+    
+    return true;
+
+}
+
+bool InverseKinematicsIPOPT::eval_jac_g(Ipopt::Index n, const Number* x, bool new_x, Ipopt::Index m, Ipopt::Index nele_jac, Ipopt::Index* iRow, Ipopt::Index* jCol, Number* values)
+{
+    
+    if(values == NULL){
+        int val = 0;
+        int i = 0;
+        int j = 0;
+    
+        //sparsity of the jacobian
+        for(val = 0; val < 7; val++){
+            iRow[val] = val; jCol[val] = val; //the top left corner of the matrix is diagonal
+        }
+        
+        for(i = 0; i < 7; i++){
+            for(j = 7; j < (totalDOF); j++){
+
+                iRow[val] = i; jCol[val] = j;  //the right top part is dense
+                val++;
+            }
+        }
+
+    }
+    else{
+        Eigen::Map< const Eigen::VectorXd > x_in(x, totalDOF);
+        Eigen::Matrix<double,Eigen::Dynamic,Eigen::Dynamic,Eigen::RowMajor> denseJac(m,totalDOF);
+        Vector4 quaternion;
+        VectorDynSize configuration;
+        MatrixFixSize<7,6> map;
+        MatrixDynSize jacobian;
+        
+        toEigen(quaternion) = x_in.segment<4>(3);
+        configuration.resize(totalDOF-7);
+        toEigen(configuration) = x_in.tail(totalDOF-7);
+        
+        twistToQuaternionTwist(quaternion, map);
+        relativeJacobian(configuration,jacobian);
+        
+        denseJac.setZero();
+        denseJac.block<7,7>(0,0) = -Eigen::Matrix<double,Eigen::Dynamic,Eigen::Dynamic,Eigen::RowMajor> (7,7).setIdentity();
+        denseJac.topRightCorner(7,totalDOF-7) = toEigen(map) * toEigen(jacobian);
+        
+        int val = 0;
+        int i = 0;
+        int j = 0;
+    
+        //sparsity of the jacobian
+        for(val = 0; val < 7; val++){
+            values[val] = denseJac(val,val); 
+        }
+        
+        for(i = 0; i < 7; i++){
+            for(j = 7; j < (totalDOF); j++){
+                values[val] = denseJac(i,j);
+                val++;
+            }
+        }
+
+    }
+    
+    return true;
+}
+
+void InverseKinematicsIPOPT::finalize_solution(SolverReturn status, Ipopt::Index n, const Number* x, const Number* z_L, const Number* z_U, Ipopt::Index m, const Number* g, const Number* lambda, Number obj_value, const IpoptData* ip_data, IpoptCalculatedQuantities* ip_cq)
+{
+    if((status == Ipopt::SUCCESS)||status == Ipopt::STOP_AT_ACCEPTABLE_POINT){
+        VectorDynSize x_in(x, totalDOF);
```

Comment:
I think the part of 'controlling' the quality of the solution can be left to the calling library.
You have obtained the best solution from the solver. Just "pack" the data and  return to the caller